src/vm/gchelpers.cpp

```diff
@@ -350,13 +350,28 @@ OBJECTREF AllocateArrayEx(TypeHandle arrayType, INT32 *pArgs, DWORD dwNumArgs, B
                 lowerBound = pArgs[i];
                 i++;
             }
+
             int length = pArgs[i];
             if (length < 0)
                 COMPlusThrow(kOverflowException);
+
             if ((SIZE_T)length > MaxArrayLength(componentSize))
+            {
+                // This will cause us to throw below if we don't throw anything else before then.
                 maxArrayDimensionLengthOverflow = true;
-            if ((length > 0) && (lowerBound + (length - 1) < lowerBound))
```

Comment:
It seems we can simplify it. If lowerBound is <= 0, we can never get overflow here. And if lowerBound and length are > 0, we can cast them to unsigned, sum them and cast them back to signed and then check if the result is negative. That should work on any compiler (I've verified that it works with clang and -O3 optimizations too).